pkg/v3/resource/namespace/resource.go

```diff
@@ -0,0 +1,119 @@
+package namespace
+
+import (
+	"context"
+	"reflect"
+
+	"github.com/giantswarm/apiextensions/pkg/apis/core/v1alpha1"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	apismetav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+
+	"github.com/giantswarm/cluster-operator/pkg/cluster"
+	"github.com/giantswarm/cluster-operator/pkg/v3/guestcluster"
+	"github.com/giantswarm/cluster-operator/pkg/v3/key"
+)
+
+const (
+	// Name is the identifier of the resource.
+	Name = "namespacev3"
+
+	namespaceName = "giantswarm"
+)
+
+// Config represents the configuration used to create a new namespace resource.
+type Config struct {
+	BaseClusterConfig        cluster.Config
+	Guest                    guestcluster.Interface
+	Logger                   micrologger.Logger
+	ProjectName              string
+	ToClusterGuestConfigFunc func(obj interface{}) (v1alpha1.ClusterGuestConfig, error)
+	ToClusterObjectMetaFunc  func(obj interface{}) (apismetav1.ObjectMeta, error)
+}
+
+// Resource implements the namespace resource.
+type Resource struct {
+	baseClusterConfig        cluster.Config
+	guest                    guestcluster.Interface
+	logger                   micrologger.Logger
+	projectName              string
+	toClusterGuestConfigFunc func(obj interface{}) (v1alpha1.ClusterGuestConfig, error)
+	toClusterObjectMetaFunc  func(obj interface{}) (apismetav1.ObjectMeta, error)
```

Comment:
So this works for all providers I've added `toClusterObjectMetaFunc` that provides access to the metadata.